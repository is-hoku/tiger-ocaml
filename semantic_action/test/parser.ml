open Lexcal_analysis
module Lexer_token = Lexcal_analysis.Token
open Semantic_action
module Parser_token = Semantic_action.Token

let tokenize buf = Parser_token.parsertoken_of_tokent (Lexer.read buf)
let parse buf = Parser.main tokenize buf
let print_ast ast = print_string (Syntax.string_of_exp ast)

let () =
  let fin = open_in "../../example_program/example3.tig" in
  let buf = Lexing.from_channel fin in
  try print_ast (parse buf) with
  | Lexer.ParseError e ->
      print_endline ("Error occurred in lexer: " ^ e);
      close_in fin
  | Lexer.SyntaxError e ->
      print_endline ("Error occurred in lexer: " ^ e);
      close_in fin
  | Parsing.Parse_error ->
      print_endline "Error occurred in parser: Parse error";
      close_in fin
