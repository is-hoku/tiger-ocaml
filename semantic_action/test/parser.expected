LetExp{decs=[VarDec{name=a, escape=true, typ=None, init=IntExp 5, pos=(2, 7)}, FunctionDec[{name=f, params=[], result=(int, (3, 19)), body=CallExp{func=g, args=[VarExp(SimpleVar(a, (3, 25)))], pos=(3, 23)}, pos=(3, 9)}, {name=g, params=[{name=i, escape=true, typ=int, pos=(4, 13)}], result=(int, (4, 25)), body=CallExp{func=f, args=[], pos=(4, 29)}, pos=(4, 9)}]], body=SeqExp[(CallExp{func=f, args=[], pos=(5, 4)}, (5, 6))], pos=(2, 3)}